### ISSUE TITLE
create event fail when working with oneToOne or morphOne relationships

### ISSUE THREAD

**Issue by CitrinoDeveloper** [wintercms/winter#413]
### Winter CMS Build

1.1

### PHP Version

7.4

### Database engine

MySQL/MariaDB

### Issue description

Hi Everyone,
I'm writing a plugin but I'm blocked on a simply one2one relationship.

In one model  User, I just want to store :
```none
----------------------------------------------------------------
id | user_unique_id | user_status | create_at | updated_at
----------------------------------------------------------------
```
Than I want to create different tables to link the user_unique_id for example Anagraphic Model:
```none
----------------------------------------------------------------
id | user_unique_id | name | surname | birthdate | ....
----------------------------------------------------------------
```

I have so created the two Models

```php
<?php namespace Citrino\Registry\Models;

use Model;
use Citrino\Registry\Models\Anagraphic as Anagraphic;

/**
 * user Model
 */
class User extends Model
{
    use \Winter\Storm\Database\Traits\Validation;

    /**
     * @var string The database table used by the model.
     */
    protected $table = 'wintertest_reserved.citrino_registry_ids';

    /**
     * @var array Relations
     */
    public $hasOne = [
           'anagraphic' => [ Anagraphic::class,'key' => 'user_unique_id', 'otherKey' => 'user_unique_id']
    ];

    protected function beforeCreate(){

        #assign a random user_unique_id
        $this->user_unique_id = random_int(0,12000000);
    }
 
}
```
And a Anagraphic Model 
```php
<?php namespace Citrino\Registry\Models;

use Model;

/**
 * registry Model
 */
class Anagraphic extends Model
{
    use \Winter\Storm\Database\Traits\Validation;

    /**
     * @var string The database table used by the model.
     */
    protected $table = 'wintertest_reserved.citrino_registry_anagraphics';

    /**
     * @var array Relations
     */
    public $belongsTo = [
        'user' => [Citrino\Registry\Models\User::class,
        'key' => 'user_unique_id']
    ];
```

Now in the user Model fields.yaml
```yaml
# ===================================
#  Form Field Definitions
# ===================================
tabs:
    fields:
        anagraphic[name]:
            tab: Anagraphic
            label: name
            type: text
            span: storm
            cssClass: col-xs-6
            required: true
        anagraphic[surname]:
            tab: Anagraphic
            label: surname
            type: text
            span: storm
            cssClass: col-xs-6
            required: true
        anagraphic[birthdate]:
            tab: Anagraphic
            label: birthdate
            type: datepicker
            mode: date
            span: storm
            cssClass: col-xs-6
        anagraphic[birthnation]:
            tab: Anagraphic
            label: nation
            type: text
            span: storm
            cssClass: col-xs-3
        anagraphic[birthplace]:
            tab: Anagraphic
            label: city
            type: text
            span: storm
            cssClass: col-xs-3
        active:
            tab: General
            label: Active
            type: checkbox
```

When creating a new item using a controller Users which point to the model User I got this message error:

```none
"Call to a member function hasRelation() on null" on line 70 of /home/rbi/rbi-he.com/current/public/modules/backend/traits/FormModelSaver.php
```

I also populated artificially the DB tables `citrino_registry_ids` and `citrino_registry_anagraphics` with dummy variable but the same `user_unique_id` and the system works perfectly and the `update` event fires without problems. 

### Steps to replicate

Create a new plugin with two models linked by a simply `one To one` relationship. 
Then try to create a new item populating the both model using a field.yaml file:

```yaml
# ===================================
#  Form Field Definitions
# ===================================
tabs:
    fields:
        anagraphic[name]:
            tab: Anagraphic
            label: name
            type: text
            span: storm
            cssClass: col-xs-6
            required: true
        anagraphic[surname]:
            tab: Anagraphic
            label: surname
            type: text
            span: storm
            cssClass: col-xs-6
            required: true
        anagraphic[birthdate]:
            tab: Anagraphic
            label: birthdate
            type: datepicker
            mode: date
            span: storm
            cssClass: col-xs-6
        anagraphic[birthnation]:
            tab: Anagraphic
            label: nation
            type: text
            span: storm
            cssClass: col-xs-3
        anagraphic[birthplace]:
            tab: Anagraphic
            label: city
            type: text
            span: storm
            cssClass: col-xs-3
        active:
            tab: General
            label: Active
            type: checkbox
```

**Comment by mjauvin** [comment]
I found the problem in modules/backend/traits/FormModelSaver.php line 74:
```
            if ($isNested && is_array($value)) {
                $this->setModelAttributes($model->{$attribute}, $value);
            }
```
It's supposed to pass a model, but is passes the relation name.

**Comment by mjauvin** [comment]
This fixes it:
```
            if ($isNested && is_array($value)) {
                $relation = $model->getRelationDefinition($attribute);
                $nestedModel = new $relation[0];
                $this->setModelAttributes($nestedModel, $value);
            }
```
Although I'm sure there's a cleaner way to do it.

**Comment by LukeTowers** [comment]
@mjauvin note that only fixes it for your specific use case. Did you test with jsonable properties or other relationship types?

**Comment by mjauvin** [comment]
It's just a starter on fixint it, we need to use deferred to assign the keys... it creates the morphed model record without linking it to the parent.

**Comment by LukeTowers** [comment]
@mjauvin for that try `$model->{$attribute}()->new()`

**Comment by mjauvin** [comment]
This works for the update context, but for create context we'd need to use a deferredBinding:
```
            if ($isNested && is_array($value)) {
                $relation = $model->getRelationDefinition($attribute);
                $nestedModel = $model->{$attribute}()->create();
                $this->setModelAttributes($nestedModel, $value);
            }
```

**Comment by mjauvin** [comment]
This will skip saving those nested fields in create context:
```
            if ($isNested && is_array($value)) {
                $context = $this->formGetContext();
                if ($context === 'create') {
                    continue;
                }
                $relation = $model->getRelationDefinition($attribute);
                $nestedModel = $model->{$attribute}()->create();
                $this->setModelAttributes($nestedModel, $value);
            }
```

**Comment by mjauvin** [comment]
And this works in both create & update context (tested for morphOne):
```
            if ($isNested && is_array($value)) {
                $context = $this->formGetContext();
                if ($context === 'create') {
                    $nestedModel = $model->{$attribute}()->create([], $this->formGetSessionKey());
                } else {
                    $nestedModel = $model->{$attribute};
                }
                $this->setModelAttributes($nestedModel, $value);
            }
```

**Comment by CitrinoDeveloper** [comment]
Dear @mjauvin,

For what I get, your code fails when hasOne relationship is used.

I have the doubt, also, that the `create()` method you are firing can skip the ` beforeCreate()` method.

Looking on the code, the iteration logic of the nesting is to recall the setModelAttributes() using a model again:

Can you (and @LukeTowers ) please give a look on the further code, and tell me if it can works?
```
  if ($isNested && is_array($value)) {
                $relationType = $model->getRelationType($attribute);
                $nestedModel = new $model->{$relationType}[$attribute][0];
                $this->setModelAttributes($nestedModel, $value);
            }
            elseif ($value !== FormField::NO_SAVE_DATA) {
                if (Str::startsWith($attribute, '_')) {
                    $attributesToPurge[] = $attribute;
                }
                $model->{$attribute} = $value;
            }
```

**Comment by mjauvin** [comment]
in your hasOne relation definition, you set both key and otherKey to the same value... that's not right

**Comment by CitrinoDeveloper** [comment]
> in your hasOne relation definition, you set both key and otherKey to the same value... that's not right

Hi @mjauvin, I understand from documentation that 'key' is my foreign key, in the example into the anagraphic table:
```
----------------------------------------------------------------
id | **user_unique_id** | name | surname | birthdate | ....
----------------------------------------------------------------
```
'otherKey' instead is the connected key into the relationship table. So, in the example into user table:
```
----------------------------------------------------------------
id | user_unique_id | user_status | create_at | updated_at
----------------------------------------------------------------
```

am I wrong with it?

**Comment by CitrinoDeveloper** [comment]
Hi @mjauvin and @LukeTowers 

on my point of view this is the last working code for this issue.


```
                if ($context === 'create') {
                    $relationType = $model->getRelationType($attribute);
                    $nestedModel = new $model->{$relationType}[$attribute][0];
                    $this->setModelAttributes($nestedModel, $value);
                    //$nestedModel = $model->{$attribute}()->create([], $this->formGetSessionKey());
                } else {
                    $nestedModel = $model->{$attribute};
                }
```

Can you check this so I can prepare a PR?

**Comment by mjauvin** [comment]
@CitrinoDeveloper this does not resolve the issue. In create context, deferredBinding is needed for the pivot table record to be created AFTER the related record is created.

**Comment by LukeTowers** [comment]
Try
```php
if ($isNested && is_array($value)) {
    // Handle related records that don't exist yet
    if (!$model->{$attribute} && $model->hasRelation($attribute)) {
        $model->{$attribute} = $model->{$attribute}()->new();
    }
    $this->setModelAttributes($model->{$attribute}, $value);
}
```

**Comment by CitrinoDeveloper** [comment]
> Try
> 
> ```
> if ($isNested && is_array($value)) {
>     // Handle related records that don't exist yet
>     if (!$model->{$attribute} && $model->hasRelation($attribute)) {
>         $model->{$attribute} = $model->{$attribute}()->new();
>     }
>     $this->setModelAttributes($model->{$attribute}, $value);
> }
> ```

"Call to undefined method Winter\Storm\Database\Relations\HasOne::new()" on line 50 of /var/www/fiorite.test/wintertest/vendor/laravel/framework/src/Illuminate/Support/Traits/ForwardsCalls.php

**Comment by mjauvin** [comment]
The following code should work for all relation types:
```
            if ($isNested && is_array($value)) {
                $context = $this->formGetContext();
                if ($context === 'create') {
                    $nestedModel = $model->{$attribute}()->create([], $this->formGetSessionKey());
                } else {
                    $nestedModel = $model->{$attribute};
                }
                $this->setModelAttributes($nestedModel, $value);
            }
```

**Comment by CitrinoDeveloper** [comment]
> The following code should work for all relation types:
> 
> ```
>             if ($isNested && is_array($value)) {
>                 $context = $this->formGetContext();
>                 if ($context === 'create') {
>                     $nestedModel = $model->{$attribute}()->create([], $this->formGetSessionKey());
>                 } else {
>                     $nestedModel = $model->{$attribute};
>                 }
>                 $this->setModelAttributes($nestedModel, $value);
>             }
> ```

As I just told before in this thread, your code jump up the `beforeCreate()` method.

**Comment by LukeTowers** [comment]
Final solution
```php
if ($isNested && is_array($value)) {
    // Handle related records that don't exist yet
    if (!$model->{$attribute} && $model->hasRelation($attribute)) {
        $model->{$attribute} = $model->{$attribute}()->getRelated();
    }
    $this->setModelAttributes($model->{$attribute}, $value);
}
```

**Comment by mjauvin** [comment]
@CitrinoDeveloper still planning on creating a PR for this?
